### PR TITLE
Remote control - display the authorization dialog in meet

### PR DESCRIPTION
**Pull request by hristoterezov**
The electron part of the PR: https://github.com/jitsi/jitsi-meet-electron/pull/10.

Depends on - https://github.com/jitsi/jitsi-meet/pull/1539

### CODE REVIEW

File: react/features/remote-control/components/RemoteControlAuthorizationDialog.js [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+     * Initializes a new RemoteControlAuthorizationDialog instance.
+     *
+     * @param {Object} props - The read-only properties with which the new
+     * instance is to be initialized.
+     */
+    constructor(props) {
+        super(props);
+
+        this._onCancel = this._onCancel.bind(this);
+        this._onSubmit = this._onSubmit.bind(this);
```
Comment (virtuacoplenny) [inline react/features/remote-control/components/RemoteControlAuthorizationDialog.js]: One edge case that bit me with device selection is it's possible to switch modals while a modal is open. So while your authorization modal is open, the user could press T and switch to speaker stats, leaving a user waiting for a response from the authorization request. I think I added a handler in componentWillUnmount but there might be a smarter fix somewhere.

File: modules/remotecontrol/Receiver.js [hunk 2]
```diff
@@ -125,52 +156,123 @@ export default class Receiver extends RemoteControlParticipant {
...
+                } else {
                     this._sendRemoteControlEvent(userId, {
                         type: EVENT_TYPES.permissions,
                         action: PERMISSIONS_ACTIONS.error
                     });
+                }
+            }).catch(() => {
+                this._sendRemoteControlEvent(userId, {
+                    type: EVENT_TYPES.permissions,
+                    action: PERMISSIONS_ACTIONS.error
```
Comment (virtuacoplenny) [inline modules/remotecontrol/Receiver.js]: One case you might want to look out for now is there is no clearRequest call in controller on permission error so the UI looks like the remote control request is pending.

File: lang/main.json [hunk 3]
```diff
@@ -331,7 +332,9 @@
         "muteParticipantTitle": "Mute this participant?",
         "muteParticipantBody": "You won't be able to unmute them, but they can unmute themselves at any time.",
         "muteParticipantButton": "Mute",
-        "remoteControlTitle": "Remote Control",
+        "remoteControlTitle": "Remote desktop control",
+        "remoteControlRequestMessage": "Would you allow __user__ to remotely control your desktop?",
```
Comment (hristoterezov) [inline lang/main.json]: @yanas Are you OK with changing this?

File: react/features/remote-control/components/RemoteControlAuthorizationDialog.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    /**
+     * Renders additional message text for the dialog.
+     *
+     * @private
+     * @returns {ReactElement}
+     */
+    _getAdditionalMessage() {
```
Comment (hristoterezov) [inline react/features/remote-control/components/RemoteControlAuthorizationDialog.js]: As far as I know our coding style is to order the methods alphabetically (constructor is exception). Currently mine are not ordered alphabetically but I'll fix this.

But I haven't heard about keeping the public methods at the top. @lyubomir what is the correct way of ordering the methods?

File: modules/API/API.js [hunk 5]
```diff
@@ -273,32 +232,17 @@ class API {
...
      *
      * @returns {void}
      */
     dispose() {
-        if (postis) {
-            postis.destroy();
-            postis = undefined;
-
+        if (this.enabled) {
+            this.enabled = false;
```
Comment (hristoterezov) [inline modules/API/API.js]: This is implemented in https://github.com/jitsi/jitsi-meet/pull/1539 . But we are doing dispose at the end which should remove all listeners.

File: modules/transport/Transport.js [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+        }
+
+        let isProcessed = false;
+
+        listenersForEvent.forEach(listener => {
+            isProcessed = listener(...args) || isProcessed;
+        });
+
+        if (!isProcessed) {
+            this._unprocessedMessages.add(args);
```
Comment (hristoterezov) [inline modules/transport/Transport.js]: This is implemented in #1539 and it is already fixed!

File: react/features/remote-control/components/RemoteControlAuthorizationDialog.js [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    /**
+     * Identify changes to the preferred input/output devices and perform
+     * necessary cleanup and requests to use those devices. Closes the modal
+     * after cleanup and device change requests complete.
+     *
+     * @private
+     * @returns {boolean} Returns false to prevent closure until cleanup is
+     * complete.
+     */
+    _onSubmit() {
```
Comment (hristoterezov) [inline react/features/remote-control/components/RemoteControlAuthorizationDialog.js]: Haha! I forgot to change the comment!

You are correct for returning true and this was my first implementation but I hit the following issue:
1. The authorization dialog opens and the user haven't started the desktop sharing
2. The user clicks Allow
3. The desktop picker dialog is supposed to open but the redux action for opening it is ignored because the authorization dialog is still displayed. 

To workaround this I had to hide the authorization dialog before dispatching the action for the desktop picker.

I'll fix the comments only.

File: conference.js [hunk 8]
```diff
@@ -1442,6 +1442,10 @@ export default {
         room.on(ConferenceEvents.DISPLAY_NAME_CHANGED, (id, displayName) => {
             const formattedDisplayName
                 = displayName.substr(0, MAX_DISPLAY_NAME_LENGTH);
+            APP.store.dispatch(participantUpdated({
+                id,
+                name: formattedDisplayName
+            }));
             APP.API.notifyDisplayNameChanged(id, formattedDisplayName);
             APP.UI.changeDisplayName(id, formattedDisplayName);
```
Comment (hristoterezov) [inline conference.js]: Good idea! 

But actually I did those changes by mistake. At the end it turned out that I don't need the display name in the store :)

I prefer to delete my display name changes than spending more time on this right now. Although  we should definitely do the change that you are proposing at some point.

File: react/features/remote-control/components/RemoteControlAuthorizationDialog.js (hunk 1 above)
Comment (hristoterezov) [inline react/features/remote-control/components/RemoteControlAuthorizationDialog.js]: Hmm. This is strange. I hit exactly the opposite problem - when a dialog is displayed, all other actions for opening another dialogs are ignored until the current dialog closes.

Are you sure that the use case that you described exists?

File: react/features/remote-control/components/RemoteControlAuthorizationDialog.js (hunk 4 above)
Comment (lyubomir) [inline react/features/remote-control/components/RemoteControlAuthorizationDialog.js]: What @virtuacoplenny said.

Comment (hristoterezov) [inline react/features/remote-control/components/RemoteControlAuthorizationDialog.js]: @lyubomir I don't mind this rule but what is the reason behind it? This brings complexity in defining the rule but what is the benefit?

File: react/features/remote-control/components/RemoteControlAuthorizationDialog.js (hunk 1 above)
Comment (virtuacoplenny) [inline react/features/remote-control/components/RemoteControlAuthorizationDialog.js]: Yes I'm sure of it. If you have the authorization dialog open and press t the modal content will switch. But maybe the fix then is to prevent other atlaskit modals from opening while one is already open, which would be for another pull request.

Comment (hristoterezov) [inline react/features/remote-control/components/RemoteControlAuthorizationDialog.js]: Yeah I think we have a lot more to do about dialogs and maybe we have issues there because with the desktop picker the behavior is exactly the opposite. And also I even cannot imaging what happens if the other dialog is from the old windows.

Also maybe we need a queue of displayed dialogs or something like that.

But let's do fix those with another PR. All those sounds like generic dialog problems.

File: lang/main.json (hunk 3 above)
Comment (yanas) [inline lang/main.json]: Sounds good!

File: react/features/remote-control/components/RemoteControlAuthorizationDialog.js [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+ *     _displayName: string
+ * }}
+ */
+function _mapStateToProps(state, ownProps) {
+    const { participantId } = ownProps;
+    const participant = getParticipantById(
+            state['features/base/participants'], participantId);
+
+    return {
+        _displayName: participant.name
```
Comment (virtuacoplenny) [inline react/features/remote-control/components/RemoteControlAuthorizationDialog.js]: A defensive check here and maybe other UI handling might be warranted for the case where the participant has left and getParticipantById returns falsy.

File: react/features/remote-control/components/RemoteControlAuthorizationDialog.js (hunk 7 above)
Comment (virtuacoplenny) [inline react/features/remote-control/components/RemoteControlAuthorizationDialog.js]: Ah, I see. receiver.grant gets called to show the desktop picker but because _onSubmit returned true, it also closes the modal that is showing desktop picker. Closing the desktop picker in this manner will also cause screensharing to no longer be toggle-able because it's the app is probably in some pending screenshare state.

File: react/features/remote-control/components/RemoteControlAuthorizationDialog.js (hunk 4 above)
Comment (hristoterezov) [inline react/features/remote-control/components/RemoteControlAuthorizationDialog.js]: I changed the order!

File: lang/main.json [hunk 10]
```diff
@@ -212,6 +212,7 @@
     },
     "dialog": {
         "add": "Add",
+        "Allow": "Allow",
```
Comment (virtuacoplenny) [inline lang/main.json]: General style nits:
- Maybe the translation key should be "allow" instead of "Allow"? The others start with a lowercase.
- Typo with "denys," should be "denies."
- Symbols in annotations are also being put in alphabetical order, so @public/@private before @returns. 
- The comment "Used to show hide the dialog on cancel." might be missing a word around "show hide." Maybe it's "show/hide." I realize it's also like this in AbstractDialog.
- You still could make actions right now for the calls to remote control within the dialog but you would leverage middleware for communicating outside of react/redux.
- I think Receiver.js's imports aren't in alphabetical order of file name. I think "/react/features" should be above "service/remotecontrol." That's the general pattern I've seen but maybe there's just inconsistency.
- Make sure to have a space after if.

File: react/features/remote-control/components/RemoteControlAuthorizationDialog.js (hunk 9 above)
Comment (hristoterezov) [inline react/features/remote-control/components/RemoteControlAuthorizationDialog.js]: How this can be possible? The participant had to be in the conference in order to send the remote control request.

Comment (virtuacoplenny) [inline react/features/remote-control/components/RemoteControlAuthorizationDialog.js]: The participant leaves the conferenece while the modal open. Maybe an alternative is passing in the display name so the component is a little dumber.

File: react/features/remote-control/components/RemoteControlAuthorizationDialog.js (hunk 7 above)
Comment (hristoterezov) [inline react/features/remote-control/components/RemoteControlAuthorizationDialog.js]: > Closing the desktop picker in this manner will also cause screensharing to no longer be toggle-able because it's the app is probably in some pending screenshare state.

I'm not sure that I'm understanding this.

File: react/features/remote-control/components/RemoteControlAuthorizationDialog.js (hunk 9 above)
Comment (hristoterezov) [inline react/features/remote-control/components/RemoteControlAuthorizationDialog.js]: OK I see now. I will fix it!

File: lang/main.json (hunk 10 above)
Comment (hristoterezov) [inline lang/main.json]: > General style nits:

> Maybe the translation key should be "allow" instead of "Allow"? The others start with a lowercase.

Fixed!

> Typo with "denys," should be "denies."

Fixed!

> Symbols in annotations are also being put in alphabetical order, so public/private before returns.

I think prams and returns should a group that shouldn't be spited.

> The comment "Used to show hide the dialog on cancel." might be missing a word around "show hide." Maybe it's "show/hide." I realize it's also like this in AbstractDialog.

Fixed!

> You still could make actions right now for the calls to remote control within the dialog but you would leverage middleware for communicating outside of react/redux.

Probably soon I'll have to refactor the remote control to move it in react. So I'll do this on this step.

> 

> I think Receiver.js's imports aren't in alphabetical order of file name. I think "/react/features" should be above "service/remotecontrol." That's the general pattern I've seen but maybe there's just inconsistency.
> 
> Make sure to have a space after if.

Will fix it!

**Comment by virtuacoplenny** [comment]
Looking good to me. Maybe you should drop the display name change commit?
